src/test/kotlin/org/jetbrains/kotlinx/jupyter/test/repl/ReplVarsSerializationTest.kt

```diff
@@ -0,0 +1,494 @@
+package org.jetbrains.kotlinx.jupyter.test.repl
+
+import io.kotest.matchers.collections.shouldContain
+import io.kotest.matchers.collections.shouldContainAll
+import io.kotest.matchers.maps.shouldContainKey
+import io.kotest.matchers.maps.shouldContainKeys
+import io.kotest.matchers.shouldBe
+import io.kotest.matchers.shouldNotBe
+import kotlinx.coroutines.runBlocking
+import org.junit.jupiter.api.Test
+
+class ReplVarsSerializationTest : AbstractSingleReplTest() {
+    override val repl = makeSimpleRepl()
+
+    @Test
+    fun simpleContainerSerialization() {
+        val res = eval(
+            """
+            val x = listOf(1, 2, 3, 4)
+            var f = 47
+            """.trimIndent(),
+            jupyterId = 1
+        )
+        val varsData = res.metadata.evaluatedVariablesState
+        varsData.size shouldBe 2
+        varsData shouldContainKey "x"
+        varsData shouldContainKey "f"
+
+        val listData = varsData["x"]!!
+        listData.isContainer shouldBe true
+        listData.fieldDescriptor.size shouldBe 2
+        val listDescriptors = listData.fieldDescriptor
+
+        listDescriptors["size"]!!.value shouldBe "4"
+        listDescriptors["size"]!!.isContainer shouldBe false
+
+        val actualContainer = listDescriptors.entries.first().value!!
+        actualContainer.fieldDescriptor.size shouldBe 2
+        actualContainer.isContainer shouldBe true
+        actualContainer.value shouldBe listOf(1, 2, 3, 4).toString().substring(1, actualContainer.value!!.length + 1)
+
+        val serializer = repl.variablesSerializer
+        serializer.doIncrementalSerialization(0, "x", "data", actualContainer)
+    }
+
+    @Test
+    fun testUnchangedVarsRedefinition() {
+        val res = eval(
+            """
+            val x = listOf(1, 2, 3, 4)
+            var f = 47
+            """.trimIndent(),
+            jupyterId = 1
+        )
+        val varsData = res.metadata.evaluatedVariablesState
+        varsData.size shouldBe 2
+        varsData.shouldContainKeys("x", "f")
+        var unchangedVariables = repl.notebook.unchangedVariables
+        unchangedVariables.isNotEmpty() shouldBe true
+
+        eval(
+            """
+            val x = listOf(1, 2, 3, 4)
+            """.trimIndent(),
+            jupyterId = 1
+        )
+        unchangedVariables = repl.notebook.unchangedVariables
+        unchangedVariables.shouldContainAll("x", "f")
+    }
+
+    @Test
+    fun moreThanDefaultDepthContainerSerialization() {
+        val res = eval(
+            """
+            val x = listOf(listOf(1), listOf(2), listOf(3), listOf(4))
+            """.trimIndent(),
+            jupyterId = 1
+        )
+        val varsData = res.metadata.evaluatedVariablesState
```

Comment:
Generally, tests are too long. I'm not sure that each test should test all internals. Each test should check one or two things